core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java

```diff
@@ -255,6 +256,11 @@ public AsyncInterceptorChain buildInterceptorChain() {
                interceptorChain.appendInterceptor(createInterceptor(new ClusteredCacheLoaderInterceptor(), ClusteredCacheLoaderInterceptor.class), false);
             else
                interceptorChain.appendInterceptor(createInterceptor(new CacheLoaderInterceptor(), CacheLoaderInterceptor.class), false);
+
+            boolean transactionalStore = configuration.persistence().stores().stream().anyMatch(StoreConfiguration::transactional);
+            if (transactionalStore && transactionMode.isTransactional())
+               interceptorChain.appendInterceptor(createInterceptor(new TransactionalStoreInterceptor(), TransactionalStoreInterceptor.class), false);
```

Comment:
Thinking about this some more, we should probably validate the configuration and not allow any store with `transactional=true` if the cache is not transactional and/or passivation is enabled.